File: 2967-minimum-cost-to-make-array-equalindromic/2967-minimum-cost-to-make-array-equalindromic.py
class Solution:
    def minimumCost(self, nums: list[int]) -> int:
        # Step 1: Sort the array to find the true median
        nums.sort()
        n = len(nums)
        median = nums[n // 2]

        # Step 2: Get the closest valid palindrome numbers surrounding the median
        lower_palindrome = self.getLowerPalindrome(median)
        upper_palindrome = self.getUpperPalindrome(median)

        # Step 3: Compute total operational cost for both candidates
        cost_with_lower = self.calculateCost(nums, lower_palindrome)
        cost_with_upper = self.calculateCost(nums, upper_palindrome)

        # Step 4: Return the minimal absolute operation cost
        return min(cost_with_lower, cost_with_upper)

    def getLowerPalindrome(self, x: int) -> int:
        while not self.isPalindrome(x):
            x -= 1
        return x

    def getUpperPalindrome(self, x: int) -> int:
        while not self.isPalindrome(x):
            x += 1
        return x

    def isPalindrome(self, x: int) -> bool:
        s = str(x)
        return s == s[::-1]

    def calculateCost(self, nums: list[int], target: int) -> int:
        cost = 0
        for num in nums:
            cost += abs(num - target)
        return cost

File: 2967-minimum-cost-to-make-array-equalindromic/2967-minimum-cost-to-make-array-equalindromic.java
import java.util.*;

class Solution {
    public long minimumCost(int[] nums) {
        // Step 1: Sort the array to find the true median
        Arrays.sort(nums);
        int n = nums.length;
        long median = nums[n / 2];

        // Step 2: Get the closest valid palindrome numbers surrounding the median
        long lowerPalindrome = getLowerPalindrome(median);
        long upperPalindrome = getUpperPalindrome(median);

        // Step 3: Compute total operational cost for both candidates
        long costWithLower = calculateCost(nums, lowerPalindrome);
        long costWithUpper = calculateCost(nums, upperPalindrome);

        // Step 4: Return the minimal absolute operation cost
        return Math.min(costWithLower, costWithUpper);
    }

    // Helper to find the closest palindrome <= x
    private long getLowerPalindrome(long x) {
        while (!isPalindrome(x)) {
            x--;
        }
        return x;
    }

    // Helper to find the closest palindrome >= x
    private long getUpperPalindrome(long x) {
        while (!isPalindrome(x)) {
            x++;
        }
        return x;
    }

    // Standard string reversal check to identify a palindrome
    private boolean isPalindrome(long x) {
        String s = Long.toString(x);
        int left = 0;
        int right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    // Accumulates absolute differences to find total operational cost
    private long calculateCost(int[] nums, long target) {
        long cost = 0;
        for (int num : nums) {
            cost += Math.abs(num - target);
        }
        return cost;
    }
}
